### DISCUSSION TITLE
关于LamdaClass starkprotocol的一些疑问

### DISCUSSION THREAD

**Discussion by readygo67** [coset-io/zkp-academy#57] [General]
1. https://lambdaclass.github.io/lambdaworks/starks/protocol.html 中红框中的话是什么意思？

<img width="763" alt="image" src="https://github.com/user-attachments/assets/ee353ad3-2d91-4763-9101-2289d2fbbe4a">

2. 在代码中没有看到红框中的这一段操作

<img width="784" alt="image" src="https://github.com/user-attachments/assets/25d3a94b-2263-4c32-a5ac-46c0164a52c2">

3. lde_periodic_columns 的作用是什么？
<img width="956" alt="image" src="https://github.com/user-attachments/assets/7319306e-f251-44e3-aafc-19789ac692bf">

4. round_2_compute_compostion_polynimal 函数中，line387 将H 分成几个部分， 在simple_fibo 中这个值是1， 即不分。在什么请况下不分，什么情况下分成多于2的部分，
<img width="1033" alt="image" src="https://github.com/user-attachments/assets/87ac3b25-1f68-412d-959a-bceff96990c1">

5. 在 https://blog.lambdaclass.com/diving-deep-fri/中有boundary constraint 和 transient constraint 对齐的操作，

![image](https://github.com/user-attachments/assets/320f5d5c-10e4-423c-9e49-6148fbee04bb)
但是在 protcol.md 中没有
![image](https://github.com/user-attachments/assets/8d2a22c2-f25e-4884-aed9-e81fe88d4ed4)
并且在代码中也没有
<img width="993" alt="image" src="https://github.com/user-attachments/assets/016c9667-1a68-46d9-914b-d1aa3d203df7">

6. 在protcol round3 中写法不够严谨 
<img width="824" alt="image" src="https://github.com/user-attachments/assets/f693c725-a69f-43a5-8eab-51e4dc3d49b4">

**Comment by Einstellung** [comment]
1. 我也不清楚，感觉不影响后续的阅读
2.  代码中有这段

```rust
    fn round_1_randomized_air_with_preprocessing(
        air: &A,
        main_trace: &TraceTable<A::Field>,
        domain: &Domain<A::Field>,
        transcript: &mut impl IsTranscript<A::FieldExtension>,
    ) -> Result<Round1<A>, ProvingError>
    where
        FieldElement<A::Field>: AsBytes + Send + Sync,
        FieldElement<A::FieldExtension>: AsBytes + Send + Sync,
    {
        // evaluations are LED evaluations of the main trace polynomials
        let (trace_polys, evaluations, main_merkle_tree, main_merkle_root) =
            Self::interpolate_and_commit::<A::Field>(main_trace, domain, transcript);

        let main = Round1CommitmentData::<A::Field> {
            trace_polys,
            lde_trace_merkle_tree: main_merkle_tree,
            lde_trace_merkle_root: main_merkle_root,
        };

        let rap_challenges = air.build_rap_challenges(transcript);
```

最后一行。

3. 我也不清楚作用是什么
4. 我想可能是trace长度比较高的时候会分一下，比如cairo的代码就分了

```rust
    fn composition_poly_degree_bound(&self) -> usize {
        2 * self.trace_length
    }
```
5. 回答在[这里](https://github.com/Antalpha-Labs/zkp-academy/discussions/56)

**Comment by keroro520** [comment]
3. Periodical constraints: A periodic constraint requires a column to have a particular value, but only at a fixed period of steps.
  在 Fibonacci 例子中没用到。貌似在 EthStark 有用到，没有深究。参考 [Periodic constraints](https://blog.lambdaclass.com/periodic-constraints-and-recursion-in-zk-starks/) 和 [examples/simple_periodic_cols.rs](https://github.com/lambdaclass/lambdaworks/blob/main/provers/stark/src/examples/simple_periodic_cols.rs) 。

**Comment by keroro520** [comment]
4. Break $CP(x)$

Composition Polynomial $CP$ 的阶可能很大。比如，对于 Trace Polynomial $f$ ，如果有个约束是 $f(x) \cdot f(x) = 0$ ，那么 $CP(x) = f(x) \cdot f(x)$ 的阶就是 $2 \times deg(f) - 1$ 。

我猜，如果所有的多项式（包括 $f$ 、$CP$ 等）的阶都在相同的 lower degree bound 内，那么对应的 Merkle Tree 和 FRI Reduction 二分的层数都相同，这一性质能让我们在工程上优化 FRI proof size。

所以，通过切分的方法让 $CP$ 的阶对齐于 Trace Polynomial $f$ 的阶。即，切分个别多项式的目的是阶的对齐。

ethSTARK 里借助 Rescue Hash Function 来描述这一优化：

![image](https://github.com/user-attachments/assets/48c79103-a366-4450-9074-b5ce66d4c88c)

**Reply by Einstellung** [reply]
确实，这样操作可以降低degree。在Cario VM中充斥着大量的f * f的约束，degree减半，可以保证CP的degree不会超过原始想承诺trace的degree。

**Reply by Einstellung** [reply]
如果degree比f * f更高，比如是Cario VM中出现过f\*f\*f的情况，此时要单独再拆出一列，始终保证f * f
```
t_0 = pc_jnz * dst
t_1 = t_0 * v
```
